IBM Granite MoE Architecture

gabe-l-hart opened the pull request:
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

## Dependencies

- [x] This PR is dependent on merging the initial `GraniteLM` PR (https://github.com/ggerganov/llama.cpp/pull/9412)

## Description

This PR introduces the `granitemoe` model architecture from IBM. It emulates the `transformers` changes in [this PR](https://github.com/huggingface/transformers/pull/33207).

The `granitemoe` architecture follows a very similar pattern to the `granite` architecture and its changes relative to `llama`. For the MoE variant, the base architecture is `mixtral` (MoE branch of `llama` here in `llama.cpp`). The same four additional multipliers are added (`embeddings_multiplier`, `attention_multiplier`, `residual_multiplier`, and `logits_scale`).

## Testing

This PR can be tested using [ibm/PowerMoE-3b](https://huggingface.co/ibm/PowerMoE-3b) from huggingface following the same testing steps used for `granite` ([here](https://github.com/ggerganov/llama.cpp/pull/9412#issue-2516885942)).

gabe-l-hart committed:
feat(gguf-py): Add granitemoe architecture

This includes the addition of new tensor names for the new moe layers.
These may not be correct at this point due to the need for the hack in
gguf_writer.py to double-check the length of the shape for these layers.

Branch: GraniteMoE

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat(convert_hf_to_gguf): Add GraniteMoeModel

GraniteMoe has the same configuration deltas as Granite

Branch: GraniteMoE

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat(granitemoe): Implement granitemoe

GraniteMoE follows the mixtral architecture (once the input_linear layers
are split into gate_exps/up_exps). The main delta is the addition of the
same four multipliers used in Granite.

Branch: GraniteMoE

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix(granitemoe convert): Split the double-sized input layer into gate and up

After a lot of staring and squinting, it's clear that the standard mixtral
expert implementation is equivalent to the vectorized parallel experts in
granite. The difference is that in granite, the w1 and w3 are concatenated
into a single tensor "input_linear." Rather than reimplementing all of the
math on the llama.cpp side, the much simpler route is to just split this
tensor during conversion and follow the standard mixtral route.

Branch: GraniteMoE

Co-Authored-By: alex.brooks@ibm.com

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -135,7 +135,7 @@ def get_total_parameter_count(self) -> tuple[int, int, int, int]:
                 size = prod(shape)
 
-                if "_exps." in name:
+                if "_exps." in name and len(shape) > 3:
```
What's the rationale behind requiring stacked expert tensor to be at least 4-dimensional?

AFAIK, these are always 3-dimensional, and so this is equivalent to `if "_exps." in name and False:`, which doesn't look correct.

gabe-l-hart replied:
Very good call! I had made this change while still trying to get things off the ground and forgot to rebate it out.

gabe-l-hart replied:
I've rebased this out now

gabe-l-hart commented:
Hi @compilade @ggerganov! This PR is now ready for full review.

We're eager to get the `granitemoe` architecture fully supported in `llama.cpp` (and then following up with [support in `ollama`](https://github.com/ollama/ollama/pull/6760#issuecomment-2362588505)). I'm sure you are perpetually swamped, so I just want to get a quick check on if this is in the review queue for you at this point and if you have any targets for merging support.

(also, thanks for the great project and all the work you do!)

ggerganov commented on convert_hf_to_gguf.py:
```diff
@@ -4102,16 +4102,43 @@ def set_gguf_parameters(self):
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        """In modeling_granitemoe, the JetMoe implementation of parallel experts
+        is used. This essentially merges w1 and w3 into a single tensor with 2x
+        the hidden size that is then split during forward. To keep compativility
```
```suggestion
        the hidden size that is then split during forward. To keep compatibility
```

gabe-l-hart committed:
Typo fix in docstring

Co-Authored-By: ggerganov@gmail.com

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>
Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
It looks like [the failing test](https://github.com/ggerganov/llama.cpp/actions/runs/10997574451/job/30533435340#step:8:2484) is on the windows server's `Erase Slot server logs` scenario. This seems like it _should_ be unrelated to this PR. Without knowing the tests well, is there any likelihood that this is a false negative? I can dig further if needed.

compilade commented on convert_hf_to_gguf.py:
```diff
@@ -4102,16 +4102,43 @@ def set_gguf_parameters(self):
+        if name.endswith("block_sparse_moe.input_linear.weight"):
+            gate, up = data_torch.chunk(2, dim=-2)
+            return [
+                (self.map_tensor_name(f"model.layers.{bid}.block_sparse_moe.input_linear.gate.weight"), gate),
+                (self.map_tensor_name(f"model.layers.{bid}.block_sparse_moe.input_linear.up.weight"), up),
```
```suggestion
                (self.format_tensor_name(gguf.MODEL_TENSOR.FFN_GATE_EXP), gate),
                (self.format_tensor_name(gguf.MODEL_TENSOR.FFN_UP_EXP), up),
```

No need to invent new intermediate names for the tensor mapping. (untested, but *should* work)

ggerganov commented:
> is there any likelihood that this is a false negative?

Yes, this is unrelated to the PR, no need to investigate.

gabe-l-hart committed:
fix(conversion): Simplify tensor name mapping in conversion

Branch: GraniteMoE

Co-Authored-By: git@compilade.net
Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

compilade commented on convert_hf_to_gguf.py:
```diff
@@ -4102,16 +4102,43 @@ def set_gguf_parameters(self):
+        """
+
+        if name.endswith("block_sparse_moe.input_linear.weight"):
+            gate, up = data_torch.chunk(2, dim=-2)
```
Since `Tensor.chunk` returns a `Tuple[Tensor, ...]`, which is not a `Tensor`, the logic for lazy evaluation can't handle this (yet), and so it's eagerly evaluated, using more memory than necessary.

Instead, slicing into the tensor should do the same thing, while being supported by the current lazy evaluation logic.

```suggestion
            ffn_dim = self.hparams["intermediate_size"]
            assert data_torch.shape[-2] // 2 == ffn_dim
            gate, up = data_torch[..., :ffn_dim, :], data_torch[..., ffn_dim:, :]
```

The size doesn't actually need to be read from `hparams`, but it can as a sanity check.

gabe-l-hart replied:
Nice, this looks like a good sanity check for sure

compilade commented on gguf-py/gguf/constants.py:
```diff
@@ -1242,6 +1244,20 @@ class MODEL_TENSOR(IntEnum):
     ],
+    MODEL_ARCH.GRANITE_MOE: [
+        MODEL_TENSOR.TOKEN_EMBD,
+        MODEL_TENSOR.OUTPUT_NORM,
```
```suggestion
        MODEL_TENSOR.OUTPUT_NORM,
        MODEL_TENSOR.OUTPUT,
```
(adding the line for `MODEL_TENSOR.OUTPUT`)
Otherwise I was getting `ValueError: Can not map tensor 'lm_head.weight'` when converting <https://huggingface.co/ibm/PowerMoE-3b>

compilade commented on src/llama.cpp:
```diff
@@ -1478,6 +1480,23 @@ static const std::map<llm_arch, std::map<llm_tensor, std::string>> LLM_TENSOR_NA
+        LLM_ARCH_GRANITE_MOE,
+        {
+            { LLM_TENSOR_TOKEN_EMBD,      "token_embd" },
+            { LLM_TENSOR_OUTPUT_NORM,     "output_norm" },
```
```suggestion
            { LLM_TENSOR_OUTPUT_NORM,     "output_norm" },
            { LLM_TENSOR_OUTPUT,          "output" },
```

Necessary for <https://huggingface.co/ibm/PowerMoE-3b>, which does have `output.weight`.

gabe-l-hart replied:
Huh! Converting `PowerMoE-3b` has worked fine for me without this. I don't think it hurts to add though and is probably a good future-proof addition for other models that use this architecture as a close cousin to `llama`.

compilade replied:
This might be because I used the weights from after <https://huggingface.co/ibm/PowerMoE-3b/commit/eac9d00a44c027335d326b62c3f45d11178220e9> which added `lm_head.weight` (aka `output.weight`), while you likely had the model downloaded before that.

gabe-l-hart replied:
Ah, spot on! Thanks for the sleuthing.

gabe-l-hart committed:
fix(convert): Remove unused tensor name mappings

Branch: GraniteMoE

Co-Authored-By: git@compilade.net
Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented on gguf-py/gguf/tensor_mapping.py:
```diff
@@ -292,10 +293,11 @@ class TensorNameMap:
+            "transformer.decoder_layer.{bid}.moe.linear_v",        # Grok (merged)
+            "transformer.blocks.{bid}.ffn.experts.mlp.v1",         # dbrx
+            "model.layers.{bid}.mlp.experts.up_proj",              # qwen2moe olmoe (merged)
+            "model.layers.{bid}.block_sparse_moe.input_linear.up", # granitemoe
```
Good call

gabe-l-hart committed:
fix(convert): Sanity check on merged FFN tensor sizes

Branch: GraniteMoE

Co-Authored-By: git@compilade.net
Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Allow "output" layer in granite moe architecture (convert and cpp)

Branch: GraniteMoE

Co-Authored-By: git@compilade.net
Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
Thanks for the detailed review @compilade! I believe I have all of the comments addressed at this point.

compilade reviewed (approved):
From the first few chunks of wikitext-2-raw with `llama-perplexity` and <https://huggingface.co/ibm/PowerMoE-3b> at `Q8_0`, I get `[1]4.4570,[2]5.1116,[3]5.3469,[4]5.9955`, so this *does* appear to work correctly.

gabe-l-hart committed:
fix(granite): Add missing 'output' tensor for Granite

This is a fix for the previous `granite` architecture PR. Recent snapshots
have included this (`lm_head.weights`) as part of the architecture

Branch: GraniteMoE

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
@compilade After you pointed out that I was missing `output` in the recent `granitemoe` snapshots, I dug a little deeper and it seems that the model team has added this for `granite` (dense) as well. I've added another commit to this PR to fix that as well. I'm not sure the preferred PR hygiene, so I'm happy to move that to a separate fix PR if the preference is for more well-encapsulated changes.
